# **ANALISIS DE LA CALIDAD DEL AIRE DE MADRID AÑO 2018**

###**CONTEXTO**

Este dataset proviene de una recopilacion de datos de monitoreo en 24 estaciones en la ciudad de Madrid, España entre los años 2001-2018. La cual surge debido a la necesidad de implementar medidas de control, como restricciones de tráfico, ante episodios de alta contaminacion dada sobre todo durante estaciones secas. Estos daros fueron proporcionados originalmente por el Ayuntamiento de dicha ciudad y han sido puesto a disposicion libre, para cualquier estudio o analisis descriptivo.

No todas las estaciones miden todos los contaminantes
Por eso existen valores nulos en varias columnas

Los principales contaminantes incluidos son:

* SO_2: Dioxido de azufre medido en μg/m³.
* NO: Oxido nitrico medido en μg/m³.
* NO_2: Dioxido de nitrogeno  μg/m³.
* PM25: Particulas en suspension de diametro menor a 2.5 μm, μg/m³.
* PM10: Particulas en suspension de diametro menor a 10 μm, μg/m³
* NOx: Oxidos de nitrosos , μg/m³.
* O_3: ozono trospoférico, μg/m³.
* TOL: tolueno (methylbenzene) ; μg/m³.
* BEN: benzeno, μg/m³.
* EBE: ethylbenzeno  μg/m³.
* MXY: m-xyleno, μg/m³.
* PXY: p-xyleno, μg/m³.
* OXY: o-xyleno, μg/m³.
* TCH: total hydrocarbonos, mg/m³.
* CH4: methano, mg/m³.
* NMHC: no-methano hydrocarbonos (volátiles organicos compouestos), mg/m³.





## **CARGA DEL DATA SET A UTILIZAR DESDE KAGGLE**

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("decide-soluciones/air-quality-madrid")

print("Path to dataset files:", path)

100%|██████████| 158M/158M [00:01<00:00, 124MB/s]

Extracting files...


Path to dataset files: /root/.cache/kagglehub/datasets/decide-soluciones/air-quality-madrid/versions/5


## **CARGA DE MODULOS REQUERIDOS EN PYTHON**

In [10]:
import pandas as pd
pd.set_option('display.float_format', '{:.2f}'.format)

In [11]:
import os

## **REVISION INICIAL DEL DIRECTORIO**

In [12]:
os.listdir(path)

['madrid.h5', 'csvs_per_year', 'stations.csv']

Se busca dentro del directorio csvs_per_year si contiene otra carpeta con el mismo nombre con los archivos csv que nos interesan

In [8]:
files = os.listdir(path + "/csvs_per_year/csvs_per_year")
files.sort()
print(files)


['madrid_2001.csv', 'madrid_2002.csv', 'madrid_2003.csv', 'madrid_2004.csv', 'madrid_2005.csv', 'madrid_2006.csv', 'madrid_2007.csv', 'madrid_2008.csv', 'madrid_2009.csv', 'madrid_2010.csv', 'madrid_2011.csv', 'madrid_2012.csv', 'madrid_2013.csv', 'madrid_2014.csv', 'madrid_2015.csv', 'madrid_2016.csv', 'madrid_2017.csv', 'madrid_2018.csv']


## **SELECCION DEL SUBCONJUNTO DEL 2018**

In [6]:
df_2018 = pd.read_csv(path + "/csvs_per_year/csvs_per_year/madrid_2018.csv", index_col='date')
df_2018.head(4)

,BEN,CH4,CO,EBE,NMHC,NO,NO_2,NOx,O_3,PM10,PM25,SO_2,TCH,TOL,station
date,,,,,,,,,,,,,,,
2018-03-01 01:00:00,NaN,NaN,0.30,NaN,NaN,1.00,29.00,31.00,NaN,NaN,NaN,2.00,NaN,NaN,28079004
2018-03-01 01:00:00,0.50,1.39,0.30,0.20,0.02,6.00,40.00,49.00,52.00,5.00,4.00,3.00,1.41,0.80,28079008
2018-03-01 01:00:00,0.40,NaN,NaN,0.20,NaN,4.00,41.00,47.00,NaN,NaN,NaN,NaN,NaN,1.10,28079011
2018-03-01 01:00:00,NaN,NaN,0.30,NaN,NaN,1.00,35.00,37.00,54.00,NaN,NaN,NaN,NaN,NaN,28079016


se guardará tambien la informacion relativa a la estaciones para una analisis complementario geoespacial, posterior.

In [15]:
df_stations = pd.read_csv(path+"/stations.csv")
df_stations.head(4)

,id,name,address,lon,lat,elevation
0,28079004,Pza. de España,Plaza de España,-3.71,40.42,635
1,28079008,Escuelas Aguirre,Entre C/ Alcalá y C/ O’ Donell,-3.68,40.42,670
2,28079011,Avda. Ramón y Cajal,Avda. Ramón y Cajal esq. C/ Príncipe de Vergara,-3.68,40.45,708
3,28079016,Arturo Soria,C/ Arturo Soria esq. C/ Vizconde de los Asilos,-3.64,40.44,693


In [17]:
df_stations["lat"].iloc[0]


np.float64(40.423852777777775)

## **ANALISIS EXPLORATORIO**

Se revisa la columna "station" para definir disponibilidad y calidad de los datos en la estaciones de monitoreo comtempladas en el estudio.

In [ ]:
df_2018["station"].nunique()

24

Se realiza un revision general de los tipos de datos en las columnas, asi como la dimension de la matriz.

In [ ]:
df_2018.info()


<class 'pandas.core.frame.DataFrame'>
Index: 69096 entries, 2018-03-01 01:00:00 to 2018-02-01 00:00:00
Data columns (total 15 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   BEN      16950 non-null  float64
 1   CH4      8440 non-null   float64
 2   CO       28598 non-null  float64
 3   EBE      16949 non-null  float64
 4   NMHC     8440 non-null   float64
 5   NO       68826 non-null  float64
 6   NO_2     68826 non-null  float64
 7   NOx      68826 non-null  float64
 8   O_3      40049 non-null  float64
 9   PM10     36911 non-null  float64
 10  PM25     18912 non-null  float64
 11  SO_2     28586 non-null  float64
 12  TCH      8440 non-null   float64
 13  TOL      16950 non-null  float64
 14  station  69096 non-null  int64  
dtypes: float64(14), int64(1)
memory usage: 8.4+ MB


In [ ]:
df_2018.shape

(69096, 15)

### **Conteo de datos nulos**

Los no nulos son:

In [ ]:
df_2018.count().sort_values()

Los nulos son:

In [ ]:
df_2018.isna().sum().sort_values()

,0
station,0
NO_2,270
NO,270
NOx,270
O_3,29047
PM10,32185
CO,40498
SO_2,40510
PM25,50184
TOL,52146


De esta fase se concluye que existen una gran cantidad de datos null para la mayoria de contaminantes observados. Por lo que para tratarlos de manera correcta se considera mejor fraccionar el analisis por categorias de estacion.

### **Revision de  estadisticas y tendencias sobre el conjunto sin filtrar.**

como columna station dice tipo int, lo cambiaremos a categoría para que no se nos muestre estadisticas sobre este campo.

In [ ]:
df_2018["station"] = df_2018["station"].astype("category")

In [ ]:
df_2018.describe()

,BEN,CH4,CO,EBE,NMHC,NO,NO_2,NOx,O_3,PM10,PM25,SO_2,TCH,TOL
count,16950.00,8440.00,28598.00,16949.00,8440.00,68826.00,68826.00,68826.00,40049.00,36911.00,18912.00,28586.00,8440.00,16950.00
mean,0.56,1.29,0.34,0.30,0.07,19.89,38.63,69.14,44.86,13.52,7.72,5.39,1.35,1.67
std,0.46,0.19,0.20,0.40,0.04,40.64,28.49,85.29,28.60,12.60,6.63,4.93,0.20,2.32
min,0.10,0.02,0.10,0.10,0.00,1.00,1.00,2.00,1.00,0.00,0.00,1.00,0.03,0.10
25%,0.30,1.14,0.20,0.10,0.04,1.00,16.00,20.00,19.00,5.00,3.00,3.00,1.20,0.40
50%,0.40,1.23,0.30,0.20,0.06,5.00,32.00,41.00,47.00,9.00,6.00,4.00,1.31,0.90
75%,0.70,1.40,0.40,0.40,0.08,18.00,55.00,84.00,68.00,18.00,11.00,7.00,1.45,2.10
max,8.40,3.92,3.20,14.90,0.49,774.00,276.00,1422.00,133.00,318.00,89.00,85.00,4.12,78.90


De este último resultado se observa que existen unas máximas fueras de escala, sobre todo en el caso de ONx (Oxidos Nitrosos).

Además del resultado de analizar los min, se notó que para Nox se consideró la suma de NO y NO2 además de algunos  oxidos nitrosos indeterminados : agrupados en NOx.

Se compararán las medias para cada uno de los gases por estacion, para inferir el comportamiento general.

Los gases como el dioxido de nitrogeno NO2 ,SO2, así como el ozone O3 (contaminacion secundaria) suelen estar asociados a transito vehicular, y a la quema de combustible fosil y no fosil como madera. Por lo que se requeriria analizar en detalle para inferir posibles causas de los picos elevados.

Se visualizan las medias por estacion:

### **Análisis Investigativo de medias y máximos por estacion**

In [ ]:
df_2018.groupby("station").mean().head(24)

Se visualizan los maximos por contaminante para cada estacion de monitoreo

In [ ]:
cols = [
    "BEN","CH4","CO","EBE","NMHC","NO","NO_2","NOx",
    "O_3","PM10","PM25","SO_2","TCH","TOL"
]

df_max = df_2018.groupby("station")[cols].max()
df_max

/tmp/ipykernel_36579/981462421.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_max = df_2018.groupby("station")[cols].max()


,BEN,CH4,CO,EBE,NMHC,NO,NO_2,NOx,O_3,PM10,PM25,SO_2,TCH,TOL
station,,,,,,,,,,,,,,
28079004,NaN,NaN,2.00,NaN,NaN,409.00,152.00,773.00,NaN,NaN,NaN,18.00,NaN,NaN
28079008,6.60,3.92,2.00,7.40,0.49,431.00,184.00,844.00,113.00,49.00,40.00,22.00,4.12,26.70
28079011,8.40,NaN,NaN,4.60,NaN,590.00,238.00,1142.00,NaN,NaN,NaN,NaN,NaN,34.80
28079016,NaN,NaN,2.40,NaN,NaN,310.00,158.00,616.00,117.00,NaN,NaN,NaN,NaN,NaN
28079017,NaN,NaN,NaN,NaN,NaN,486.00,181.00,879.00,133.00,NaN,NaN,20.00,NaN,NaN
28079018,6.30,NaN,2.10,14.90,NaN,350.00,185.00,714.00,119.00,318.00,NaN,25.00,NaN,49.20
28079024,3.30,3.33,0.80,6.30,0.24,162.00,130.00,361.00,129.00,77.00,42.00,14.00,3.47,14.80
28079027,NaN,NaN,NaN,NaN,NaN,308.00,169.00,640.00,119.00,NaN,NaN,NaN,NaN,NaN
28079035,NaN,NaN,1.40,NaN,NaN,311.00,118.00,566.00,80.00,NaN,NaN,19.00,NaN,NaN


Se convierte a formato largo para visualizar los datos maximos por estacion y por contaminante

---



In [ ]:
df_long = df_max.stack().reset_index()
df_long.columns = ["station", "contaminante", "valor"]
df_long

### **Ranking global de maximos :**

Con este paso se logra determinar las estaciones  donde ocurrieron los máximos globales de cada contaminante.

In [ ]:
df_max_contaminante = df_long.loc[
    df_long.groupby("contaminante")["valor"].idxmax()
]
df_max_contaminante

,station,contaminante,valor
19,28079011,BEN,8.40
6,28079008,CH4,3.92
84,28079039,CO,3.20
37,28079018,EBE,14.90
9,28079008,NMHC,0.49
85,28079039,NO,774.00
129,28079056,NO_2,276.00
87,28079039,NOx,1422.00
33,28079017,O_3,133.00
42,28079018,PM10,318.00


## **CRUCE DE INFORMACION DE MAXIMOS CON LA FECHA DEL ELVENTO**

In [ ]:
df_2018 = df_2018.reset_index()
cols = [
    "BEN","CH4","CO","EBE","NMHC","NO","NO_2","NOx",
    "O_3","PM10","PM25","SO_2","TCH","TOL"
]

resultados = []

for col in cols:
    idx = df_2018[col].idxmax()
    fila = df_2018.loc[idx]

    resultados.append({
        "contaminante": col,
        "station": fila["station"],
        "fecha": fila["date"],
        "valor": fila[col]
    })

df_eventos = pd.DataFrame(resultados)


In [ ]:
df_eventos.sort_values("fecha")

,contaminante,station,fecha,valor
9,PM10,28079018,2018-01-01 02:00:00,318.00
13,TOL,28079055,2018-01-22 17:00:00,78.90
0,BEN,28079011,2018-01-23 22:00:00,8.40
2,CO,28079039,2018-01-23 22:00:00,3.20
4,NMHC,28079008,2018-01-23 22:00:00,0.49
5,NO,28079039,2018-01-23 22:00:00,774.00
7,NOx,28079039,2018-01-23 22:00:00,1422.00
6,NO_2,28079056,2018-01-24 21:00:00,276.00
10,PM25,28079056,2018-01-24 22:00:00,89.00
11,SO_2,28079036,2018-01-28 21:00:00,85.00


Del resultado de eventos máximos se puede observar, que las maximas para PM10 y PM2,5 ocurren en disntintas fechas y horarios nocturnos, aunque puede ser que para una misma fecha ambos contaminantes hayan bordeado el maximo que se indica.Se revisará más adelante.

Así mismo para  BEN, CO, NMHC se ovserva que ocurren en la misma fecha y hora , 00:00, econtrandose tambien el max de TOL en un día anterior y 5 horas antes de las 22:00, lo cual sugiere que para la misma fecha TOL podría haber bordeado su máximo.

Para los oxidos Nitrosos, NO, NO2 Y NOx, se observa que se comportaron similar alcanzadno maximo en la misma fecha y hora pra el caso de NO y NOx, las cuales además pertenecen a la misma estacion. En el caso de NO2 el maximo se confirmo en la estacion "056" un día  anterior y a la misma hora que los otros 2 gases. De la tabla general de datos se confirma que para la misma fecha los 3 gases bordearon los maximos indicados. Esto indica claramente que este evento está relacionado a una misma causa.

para el caso de O_3 , ozone, se observa que su máximo se registra a las 02:00 horas a fines de Abril. Muy distante de la maxima de CO que se registra a fines de Enero. No obstante,sabiendo que CO es precursor de O_3 por ser catalizador químico, es posible para una misma fecha Los valores de ambos bordeen sus maximos respectivos. Esto se confirma en la tabla de maximos globales, en donde para la estacion que registra el maxima de CO (3,2) el valor de ozone observado por la misma estacion es de 117, elevado, aunque no tan cerano a su maxima de 133 registrada en abril. El caso opuesto de la tabla global mencionada se observa que en la estacion donde se registra la máxima para ozone no se dispone de informacion para CO.

Para continuar el analisis filtraremos por fecha y hora, a fín de inferir si los picos pueden estar relacionados a un mismo evento, lo cual se complementará con investigación bibliografica. Este paso se realizará usando power Bi para complementar más de un tipo de herramienta de analisis y visualización.

## **NORMALIZACION DEL DATA SET DE CALIDAD DL AIRE, Z-SCORE**

Se realiza una normalizacion de los datos, teniendo en cuenta cada columna de contaminante así como su estación de monitoreo. De esta manera es mas fiable la comparación entre columnas.Se aplicará una normalizacion Z-score.

In [ ]:
df_2018 = df_2018.drop(columns=["level_0", "index"], errors="ignore")
cols = [
    "BEN","CH4","CO","EBE","NMHC","NO","NO_2","NOx",
    "O_3","PM10","PM25","SO_2","TCH","TOL"
]
df_2018[cols] = df_2018.groupby("station")[cols].transform(
    lambda x: (x - x.mean()) / x.std()
)

/tmp/ipykernel_36579/3421038659.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_2018[cols] = df_2018.groupby("station")[cols].transform(


In [ ]:
df_2018

,date,BEN,CH4,CO,EBE,NMHC,NO,NO_2,NOx,O_3,PM10,PM25,SO_2,TCH,TOL,station
0,2018-03-01 01:00:00,NaN,NaN,-0.66,NaN,NaN,-0.47,-0.47,-0.50,NaN,NaN,NaN,-0.26,NaN,NaN,28079004
1,2018-03-01 01:00:00,-0.61,-0.44,-0.53,-0.40,-0.79,-0.58,-0.64,-0.63,0.59,-0.73,-0.60,-0.42,-0.57,-0.61,28079008
2,2018-03-01 01:00:00,-0.12,NaN,NaN,0.16,NaN,-0.44,-0.08,-0.34,NaN,NaN,NaN,NaN,NaN,0.01,28079011
3,2018-03-01 01:00:00,NaN,NaN,-0.24,NaN,NaN,-0.46,-0.19,-0.39,0.41,NaN,NaN,NaN,NaN,NaN,28079016
4,2018-03-01 01:00:00,NaN,NaN,NaN,NaN,NaN,-0.46,-0.43,-0.47,0.31,NaN,NaN,-0.64,NaN,NaN,28079017
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
69091,2018-02-01 00:00:00,NaN,NaN,0.68,NaN,NaN,0.40,1.05,0.60,-1.35,1.05,1.29,NaN,NaN,NaN,28079056
69092,2018-02-01 00:00:00,NaN,NaN,2.43,NaN,NaN,2.03,2.58,2.36,NaN,1.16,NaN,2.48,NaN,NaN,28079057
69093,2018-02-01 00:00:00,NaN,NaN,NaN,NaN,NaN,2.31,2.28,2.53,-1.63,NaN,NaN,NaN,NaN,NaN,28079058
69094,2018-02-01 00:00:00,NaN,NaN,NaN,NaN,NaN,4.59,3.17,4.39,-1.62,NaN,NaN,NaN,NaN,NaN,28079059


Tal como se habia identificado antes, existen además de picos detectados anteriormente,en el registro del n°69092 al n°69095, final de la tabla, varios outliers para el CO y los NOx en general.

In [ ]:
df_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69096 entries, 0 to 69095
Data columns (total 16 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   date     69096 non-null  object  
 1   BEN      16950 non-null  float64 
 2   CH4      8440 non-null   float64 
 3   CO       28598 non-null  float64 
 4   EBE      16949 non-null  float64 
 5   NMHC     8440 non-null   float64 
 6   NO       68826 non-null  float64 
 7   NO_2     68826 non-null  float64 
 8   NOx      68826 non-null  float64 
 9   O_3      40049 non-null  float64 
 10  PM10     36911 non-null  float64 
 11  PM25     18912 non-null  float64 
 12  SO_2     28586 non-null  float64 
 13  TCH      8440 non-null   float64 
 14  TOL      16950 non-null  float64 
 15  station  69096 non-null  category
dtypes: category(1), float64(14), object(1)
memory usage: 8.0+ MB


##**EXPORTACION A EXCEL Y CSV PARA POSTERIOR ANALISIS EN POWER BI**


In [19]:
os.chdir('/content/drive/MyDrive/DATA_SET_PROJECTS/')

In [ ]:
df_2018["date"] = pd.to_datetime(df_2018["date"])

In [ ]:
df_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69096 entries, 0 to 69095
Data columns (total 16 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   date     69096 non-null  datetime64[ns]
 1   BEN      16950 non-null  float64       
 2   CH4      8440 non-null   float64       
 3   CO       28598 non-null  float64       
 4   EBE      16949 non-null  float64       
 5   NMHC     8440 non-null   float64       
 6   NO       68826 non-null  float64       
 7   NO_2     68826 non-null  float64       
 8   NOx      68826 non-null  float64       
 9   O_3      40049 non-null  float64       
 10  PM10     36911 non-null  float64       
 11  PM25     18912 non-null  float64       
 12  SO_2     28586 non-null  float64       
 13  TCH      8440 non-null   float64       
 14  TOL      16950 non-null  float64       
 15  station  69096 non-null  category      
dtypes: category(1), datetime64[ns](1), float64(14)
memory usage: 8.0 MB


In [ ]:
df_normalizado = df_2018.copy()


In [ ]:
df_normalizado.to_excel("air_quality_normalized.xlsx", index=False)

In [ ]:
df_max.to_csv("eventos_extemos.csv", index=False)

In [20]:
df_stations.to_csv("Estaciones_georeferenciadas.csv", index=False)